### ISSUE TITLE
Extended statistics

### ISSUE THREAD

**Issue by rtaupe** [alpha-asp/Alpha#113]
Currently, Alpha is able to log a small number of statistical values on the solving process (cf. #51, #99, #102).
To gather more data, it could be beneficial to log other values as well. **Which ones should that be?**

For example, clasp outputs something like the following:
> Models       : 1+
> Calls        : 1
> Time         : 28.071s (Solving: 27.05s 1st Model: 27.04s Unsat: 0.00s)
> CPU Time     : 27.191s
> 
> Choices      : 133724
> Conflicts    : 103856   (Analyzed: 103856)
> Restarts     : 331      (Average: 313.76 Last: 167)
> Model-Level  : 26.0
> Problems     : 1        (Average Length: 0.00 Splits: 0)
>
> Lemmas       : 103856   (Deleted: 77119)
>>   Binary     : 1655     (Ratio:   1.59%)
>>   Ternary    : 2541     (Ratio:   2.45%)
>>   Conflict   : 103856   (Average Length:   33.4 Ratio: 100.00%)
>>   Loop       : 0        (Average Length:    0.0 Ratio:   0.00%)
>>   Other      : 0        (Average Length:    0.0 Ratio:   0.00%)
>
> Backjumps    : 103856   (Average:  1.26 Max:  48 Sum: 130701)
>>   Executed   : 103856   (Average:  1.26 Max:  48 Sum: 130701 Ratio: 100.00%)
>>   Bounded    : 0        (Average:  0.00 Max:   0 Sum:      0 Ratio:   0.00%)
> 
> Rules        : 105814
> Atoms        : 36917
> Bodies       : 67239
> Equivalences : 37788    (Atom=Atom: 303 Body=Body: 1758 Other: 35727)
> Tight        : Yes
> Variables    : 22212    (Eliminated:    0 Frozen:    0)
> Constraints  : 85168    (Binary:  77.7% Ternary:  22.2% Other:   0.1%)

**Comment by AntoniusW** [comment]
Those statistics are indeed interesting. In principle, we could add all those statistics, although several of them are meaningless at the moment (e.g., Alpha currently does not delete Lemmas, we have no tight-ness analysis, and no variable elimination).

**Comment by peschue** [comment]
As alpha works very different from gringo+clasp, some statistics will make sense in a more extended version.

I imagine it is interesting to count rules/atoms/bodies up to the first answer set and then separately from the first to the second and so on ...

I imagine statistics about how often each case in the main loop of the Alpha algorithm (Algorithm 2 in LPNMR paper) is hit would be interesting and how much time is spent in each.

Another number I would look at is the size of the set returned by lazyGround (Algorithm 2 in LPNMR paper) and the distribution of the size of the set.

Such benchmarking will influence the performance, in particular caching. Maybe parameterizing the algorithm with a statistics gathering object (I think this is possible in Java) and compiling two versions into the solver that can be used interchangeably (only one collects a huge amount of statistics) is possible. This way one JAR file could either run fast without statistics and slower with statistics.

**Comment by AntoniusW** [comment]
I am not fully sure what the counting of rules/atoms/bodies from one answer-set to another is really telling us? One probably can observe whether between two answer-sets some large space had to be grounded/explored. Are there other conclusions we can draw from having these specific numbers?

Other than that, I fully agree with @peschue.

Regarding "compiling two versions": I think this can be relayed to the JIT compiler of the JVM. That is, we only need to compile one version and make it visible to the JIT compiler that the statistics gathering is enabled by some kind of global flag that won't be turned on later when it is off in the beginning. In principle, the JIT compiler should then remove all statistics-gathering code from the JIT-compiled code directly.

**Comment by rtaupe** [comment]
I fully agree with @peschue. As long as it does not hurt to gather more data, we may find them useful when we have them.

@AntoniusW you are of course right, it is not desirable to compile two different versions. It should be possible to switch gathering of statistics on and off via command-line flags. Do you think what you are suggesting is possible with that? Do you have a reference describing this technique?

**Comment by peschue** [comment]
The method I sketched would prevent to compile two versions, it would compile the same class two times with different template parameters and build one executable that can run in two modes.

Of course if JIT can solve this transparently, that would be a much better method!

**Comment by AntoniusW** [comment]
Unfortunately, I didn't find a documentation of specific JIT techniques for that. The JIT does remove dead code, but to what extent is not fully clear. Furthermore, if the statistics-gathering is guarded by a simple if-check, then branch-prediction of every modern CPU will lead to near-zero cost. So I guess, what we are talking is actually premature optimization. Just wrap all statistics inside an `if (statisticsOn) { ... }` should be fine.

@peschue by templates you mean Generics that look like templates in C++, right? Generics in Java are completely removed by the compiler by a process called Type Erasure. In the end there is only a single class without any parameters remaining.

**Comment by lorenzleutgeb** [comment]
Regarding collection of statistics: I think using the JIT as a "feature" when designing a system is suboptimal.

My approach would be as follows.

 * Define an interface of what counters exist and name them. For each counter, add a method to the interface that can be used to read the counter (alternatively one to read all counters) and one to increment it or do some tricky bookkeeping and/or math.
 * Implement the interface so that it actually counts.
 * Implement the interface with empty method bodies and name it accordingly.

```java
interface StatisticsCollector {
    public void countBacktrack();
    public void reportLazyGroundSize();

    public void printReport();
}

class NullStatisticsCollector {
    public void countBacktrack() { }
    public void reportLazyGroundSize() { }

    public void printReport() { println("NO STATISTICS COLLECTED"); }
}

class StandardStatisticsCollector {
    public void countBacktrack() { ... }
    public void reportLazyGroundSize() { ... }

    public void printReport() { println("..."); }
}
```

 * The JIT will have an easy life optimizing away empty methods.
 * No `if` clauses to check whether statistics are enabled that litter our code and make it had to follow. A call to `stat.foo()` is very expressive.
 * A `StatisticsCollector` instance can be passed around to all components that want to record statistics. It would tickle down in initialization similar to `Random` instances seeded via the CLI.
 * If we ever do parallel computations, we may have n different `StatisticsCollector` instances and could define how to merge them.

**Comment by rtaupe** [comment]
Statistics on numbers of atoms should distinguish between different types of atoms (`BasicAtom`, `ChoiceAtom`, `RuleAtom`, ...)

**Comment by rtaupe** [comment]
Problem in last comment is addressed by d60762172302af9baea0864590f3a801116f2829

**Comment by rtaupe** [comment]
Note: Performance statistics introduced by #137 are reset every time we enter `DefaultSolver.tryAdvance(...)`, i.e. if we search for more than one answer set, it is reset every time an answer set is found. I think this should be changed in a future fix.

**Comment by rtaupe** [comment]
> 
> 
> Note: Performance statistics introduced by #137 are reset every time we enter `DefaultSolver.tryAdvance(...)`, i.e. if we search for more than one answer set, it is reset every time an answer set is found. I think this should be changed in a future fix.

Note that only the time is reset after an answer set has been found, not the number of decisions. This means that the overall number of decisions is divided by the time that has passed since the last answer set has been found, which does not make sense.

Opened new issue #166 for this.
